File: 2020/7.php
<?php

require('libs/core.php');

function can_contain($bag, $bags) {
    if (array_key_exists("shiny gold", $bag)) {
        return true;
    }
    foreach (array_keys($bag) as $colour) {
        if (isset($bags[$colour]) && can_contain($bags[$colour], $bags) === true) {
            return true;
        }
    }
}

function total_bags($bag, $bags) {
    $t = 0;
    foreach($bag as $colour => $number) {
        if (isset($bags[$colour])) {
            $t += $number + ($number * total_bags($bags[$colour], $bags));
        }
    }
    return $t;
}

function check_1($bags) {
    $total = 0;
    foreach ($bags as $contains) {
        $can = can_contain($contains, $bags);
        if ($can) {
            $total++;
        }
    }
    print "{$total} matching\n";
}

function check_2($bags) {
    $total = total_bags($bags["shiny gold"], $bags);
    print "{$total} in total\n";
}

$data = load_data("7.txt");

$bags = [];
foreach ($data as $rule) {
    if (preg_match_all('/(?:(\d) )*(\w+ \w+) bags*/', $rule, $matches)) {
        print_r($matches);
        $bags[$matches[2][0]] = [];
        for ($i = 1; $i < count($matches[2]); $i++) {
            $bags[$matches[2][0]][$matches[2][$i]] = $matches[1][$i] ?: 0;
        }
    }
}

check_1($bags);

check_2($bags);

File: 2020/python/7.py
import core
import re

def get_bags(data):
    bags = {}
    for rule in data:
        matches = re.findall('(?:(\d) )*(\w+ \w+) bags*', rule)
        contains = {}
        for i in range(1, len(matches)):
            contains[matches[i][1]] = int(matches[i][0] or 0)
        bags[matches[0][1]] = contains
    return bags

def can_contain(bag, bags):
    if "shiny gold" in bag:
        return True
    for child in bag:
        if (child in bags) and can_contain(bags[child], bags):
            return True

def total_bags(bag, bags):
    total = 0;
    for child in bag:
        if child in bags:
            number = bag[child]
            total += number + (number * total_bags(bags[child], bags))
    return total;

def check_1(bags):
    total = 0;
    for colour in bags:
        can = can_contain(bags[colour], bags)
        if can:
            total += 1
    print("%d matching" % total)

def check_2(bags):
    total = total_bags(bags["shiny gold"], bags);
    print("%d in total" % total);

data = core.load_data("7.txt")

bags = get_bags(data);

check_1(bags)

check_2(bags);
